spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/procedures/procedure.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.procedures
+
+import org.apache.spark.sql.execution.command.RunnableCommand
+import org.apache.spark.sql.types.DataType
+
+trait Procedure {
+  def name: String
+  def parameters: Seq[Parameter]
+  def createRunnableCommand(args: Array[Any]): RunnableCommand
```

Comment:
Using `RunnableCommand` avoids the need to create a physical plan, right? Is it possible to change to using [`V2CommandExec`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2CommandExec.scala) instead?

That way, commands can expose their physical plans and we don't end up creating a physical plan inside a logical plan, which leads to weird issues.

Reply:
`RunnableCommand` is a logical plan that gets wrapped into `ExecutedCommandExec` by a strategy. I used it because it does not need an extra logical plan node like `CallProcedure` and an extra strategy to convert into `V2CommandExec`. However, we can reconsider that.

Reply:
Yes, but the problem with `RunnableCommand` is that it is a logical plan. Executing it with `ExecutedCommandExec` runs a logical plan in your physical plan. That's okay if it just needs to do one simple thing. But if it needs to do something that _should_ be done by a query it is a bad practice. For all of the v2 plans, we opted to use real physical plans to avoid problems.

Reply:
I updated it to be `CallStatement` -> `Call` -> `CallExec`. Also, `V2CommandExec` requires `InternalRow`s so I added conversion for that too.